Move location fields out of user data

esoergel opened the pull request:
## Product Description
<!-- Where applicable, describe user-facing effects and include screenshots. -->

## Technical Summary
https://dimagi.atlassian.net/browse/USH-4428
This is an alternative approach to https://github.com/dimagi/commcare-hq/pull/35292 I move location data out of user data entirely, injecting it at the last minute.  We already did this for web users as of https://github.com/dimagi/commcare-hq/pull/34401, but there, the fields were _always_ present, whereas for mobile workers, they're only present if not null.  This commit brings web users in line with the behavior for mobile workers, for better parity

-----

Custom user data is an unschema'd key/value mapping where projects can throw whatever information they want to associate with their users.  This is made available in the restore file so it can be referenced in apps.  At some point in Dimagi's history, we wanted to make the user's assigned location also available for reference in the restore file.  This should have been implemented by adding it (along with other system fields) to the session data at the last minute.  Instead, we wrote handlers for setting, unsetting, and appending the user's locations, and made those mutate the user_data whenever the main location fields were modified, duplicating the information there.

This was tedious to deal with, but it also had the consequence of making these fields available anywhere user data is available, unless someone specifically exempted them.  User data is referenced in a number of places:

* `get_user_session_data` - this is pulled into the registration element in the restore.  That's the main reason we have this data - so it can be referenced in commcare apps.  Other system fields are pulled in here as well - first and last name, phone number, and user_type.
* Edit user UI - "system" fields are hidden
* User import/export - "system" fields are hidden
* API - this gets a pretty raw view of the user doc, including all user data, though system fields don't really belong here, since they are all more readily available in other parts of the doc.  If I weren't concerned about changing user-facing behavior, I'd remove these location fields entirely.  We had a bug report somewhat recently from someone trying to set user locations by modifying these fields in the API, but that's not at all how these work.
* UCR - these fields are likewise available for reference in UCRs, though they are better referenced directly from the top level, and doing so will let location filters work properly, and allow these reports to be made available to location-restricted users.
* `_get_user_case_fields` - this makes data available on the usercase.  If it wouldn't break existing apps, here I'd drop `commcare_primary_case_sharing_id` and make the other two in-line with the other system fields being added just below there (make the fields always present, even when null, and drop the `commcare_` prefix)

Making changes to user-facing behavior in any of these places has the potential to disrupt existing workflows.  In this PR I've gone the route of decoupling and localizing the implementations, since there is no real conceptual linkage between them, besides an accident of history.  My hope is that we stop perpetuating the problem (that changing one changes all of them), and make it easier to correct specific implementations as they go through future changes.  For example, maybe the next version of the API drops them entirely, or we audit usercase usage before GA and decide to bring the fields in line with the other system fields.

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->


### Migrations
<!-- Delete this section if the PR does not contain any migrations -->
<!-- https://commcare-hq.readthedocs.io/migrations_in_practice.html -->
- [ ] The migrations in this code can be safely applied first independently of the code

<!-- Please link to any past code changes that are coordinated with this migration -->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [ ] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [ ] Risk label is set correctly
- [ ] The set of people pinged as reviewers is appropriate for the level of risk of the change

esoergel committed:
Move location fields out of user data

This is an alternative approach to https://github.com/dimagi/commcare-hq/pull/35292
I move location data out of user data entirely, injecting it at the
last minute.  We already did this for web users as of
https://github.com/dimagi/commcare-hq/pull/34401, but there, the fields
were _always_ present, whereas for mobile workers, they're only present
if not null.  This commit brings web users in line with the behavior for
mobile workers, for better parity

esoergel committed:
Move commcare_project out too

mjriley commented:
>  In this PR I've gone the route of decoupling and localizing the implementations, since there is no real conceptual linkage between them, besides an accident of history. My hope is that we stop perpetuating the problem (that changing one changes all of them), and make it easier to correct specific implementations as they go through future changes.

Do we foresee that different implementations will require different parts of location data? I can understand not wanting to glob location data into user data all the time, as like you pointed out, it duplicates the 'source-of-truth' data and then encourages users to use that duplicated data rather than the original. However, it still seems like all of these implementations just want to expose 'location data', and 'location data' means the same thing for every implementation. Would it make sense to extract out the `commcare_location_id`, `commcare_location_ids`, and `commcare_primary_case_sharing_id` fields into a separate helper function like `get_location_data()` that returns a dict of these values? This would convey the logical grouping, allow the implementation to modify or ignore fields if necessary, and remove the duplication.

esoergel committed:
Add location ID back to UCR data sources

A few projects do use this field, see for details:
https://docs.google.com/document/d/1MM1lOnFArfr8F6CbgQ9OvvzpWBlwZJEB780YFA7NGls/edit?tab=t.0

esoergel commented:
> Do we foresee that different implementations will require different parts of location data?

Yes.  I'm pretty sure nothing actually needs `commcare_primary_case_sharing_id`, for one.  Then after that, it's two `CommCareUser` fields among many, one an ID, the other a list of IDs.  SolTech recently audited UCR fields and from the artifacts of that I can see that there are a handful of projects which reference `commcare_location_id`, so that can't easily be removed from that instance, but the other fields can be (just pushed a commit to that effect).  The restore data and usercase data both need the two location fields, but the names _should_ be different (system fields are all prefixed with `commcare_` in the former, but not in the latter, except for these ones).  The API should present the list of IDs as a JSON array rather than a space-separated list (and these fields shouldn't be present in user data at all).  The implementations are the same in this PR because I don't want to introduce behavior changes at this time, but I'd actually like to encourage them to drift apart from one another and more towards the other system fields in their respective contexts.

esoergel committed:
update tests

esoergel committed:
Use modern string formatting

The positional arguments of the first fn in particular are brittle
